### ISSUE TITLE
Changing default distance metric for misorientations

### ISSUE THREAD

**Issue by pc494** [pyxem/orix#199]
The discussion on #196 suggests we would ideally use the misorientation angle (over the geodesic distance between quaternions) but it proved troublesome to implement. Revisiting this at some point in the future would be good.

**Comment by hakonanes** [comment]
I think this is an important feature and we should have a memory and time efficient to calculate misorientations, both between orientations with the same symmetry and with different symmetry. I think this should include the use of at least Numba, but also possibly Dask.

I just found out that `Orientation.get_distance_matrix()` doesn't play well with 2D arrays (n, abcd). It was written using Dask to enable calculating of the distance matrices in the Applied Crystallography paper, where `Orientation` in the user guide has a shape of (100, 200). For this it works fine.

I would like to bring #212 in here and close that issue, so we can coordinate discussion:

> @harripj's addition of transposing orientation's navigation dimensions should be looked at as a potential speed up of calculations misorientation angles. See #209.

I don't have time to work on this myself, but can help anyone who has.

**Comment by harripj** [comment]
@hakonanes thanks for the ping. I was using the transpose in a specific case where I had ordered data between two sets of Orientations, but this is not the case in general I suppose. It would be good to have a common way to calculate the misorientation angles, and it seems to me that ```Orientation.get_distance_matrix()``` is a special case of this where the Orientations are compared with themselves. I would argue in this case that the function should fall back to the same calculation and so should work ndim.

Please correct me where I am mistaken here as I am not 100% familiar with the code.

**Comment by hakonanes** [comment]
> I was using the transpose in a specific case where I had ordered data between two sets of Orientations, but this is not the case in general I suppose.

Could I ask how many orientations you had?

> It would be good to have a common way to calculate the misorientation angles 

I agree, that would be best. In general just a simple non-vectorized function which takes in two orientations with symmetries set, and returns the misorientation angle. If we could agree upon such a function, that would make making vectorized functions simpler, since we could check against this non-optimized but correct function.

> it seems to me that `Orientation.get_distance_matrix()` is a special case of this where the Orientations are compared with themselves. I would argue in this case that the function should fall back to the same calculation and so should work ndim.

I agree, but `get_distance_matrix()` was written specifically to avoid going out of memory when computing misorientation angles with a large `Orientation` instance by using Dask. But as I said there must be some assumptions on Orientation.shape that doesn't play well with 1D Orientation instance.

Again, having a "ground truth" function to test against would make having multiple implementations for different circumstances not a problem, in my opinion.

It's worth noting that MTEX' own Orientation/angle function also fills memory, so you would have to loop in MTEX.

> Please correct me where I am mistaken here as I am not 100% familiar with the code.

I'm not here to correct anyone, thanks for taking part in the discussion!

**Comment by harripj** [comment]
> Could I ask how many orientations you had?

It was only a few hundred, not really a problem for looping, but I thought it best to run it vectorised in any case as it was simple to implement.

> I agree, that would be best. In general just a simple non-vectorized function which takes in two orientations with symmetries set, and returns the misorientation angle. If we could agree upon such a function, that would make making vectorized functions simpler, since we could check against this non-optimized but correct function.

I'm just going to throw this here to start off a discussion:
```python
from orix.quaternion import Orientation
from orix.quaternion.symmetry import Oh, C1

o1 = Orientation.random_vonmises()
o2 = Orientation.random_vonmises()

o1 = o1.set_symmetry(Oh)

def single_disorientation(o1, o2):
    '''Disorientation between two orientations.'''
    assert len(o1.shape) == 1 and o1.shape[0] == 1, 'o1 must be a single orientation.'
    assert len(o2.shape) == 1 and o2.shape[0] == 1, 'o2 must be a single orientation.'

    # apply symmetry
    o10 = o1.outer(o1.symmetry)[0]  # [0] to just get symmetry orientations
    o20 = o2.outer(o2.symmetry)[0]

    misorientation = o10.outer(~o20)
    disorientation = misorientation.angle.data.min()
    return disorientation
    
single_disorientation(o1, o2)
>>> 0.67  # for me, your mileage will vary
```

> Again, having a "ground truth" function to test against would make having multiple implementations for different circumstances not a problem, in my opinion.

Completely agree.

**Comment by hakonanes** [comment]
Thanks! This function looks correct to me, although it would be good to get @pc494's view on it when he finds the time.

**Comment by pc494** [comment]
Function looks correct to me, `# [0] to just get symmetry orientations` confused me but otherwise this forms a great place to start from!

**Comment by harripj** [comment]
@pc494 yes you're right, possibly not the most explicit comment! The ```[0]``` is there to index the first (and only) navigation dimension for the ```Orientation```s (of which there is only 1 of length 1 due to the ```assert``` statements). In principle this is not necessary as the minimum angle is taken over all the navigation dimensions, but just there to simplify the dimensionality of ```misorientation``` and to make clear that we are only considering single ```Orientation```s.

**Comment by pc494** [comment]
I ran it in an ipython window and it made sense, then I really struggled to find a better wording for the comment...

**Comment by hakonanes** [comment]
I consider this done since both `Misorientation` and `Orientation` now have `get_distance_matrix()` (see docs for [Misorientation](https://orix.readthedocs.io/en/latest/reference.html#orix.quaternion.Misorientation.get_distance_matrix)). What we don't have is `Misorientation.angle_with()`, but if that becomes necessary, another issue should be created.
